Allow cfg attribute macros on metadata fields

oppiliappan opened the pull request:
This PR should fix #381.

TODO:

 - [x] add tests (I am not sure how this is done)
 ~emit correct error message on duplicate field declaration~

oppiliappan committed:
rework rate_limited and authentication to include attributes

oppiliappan committed:
limit to cfg attributes

oppiliappan committed:
Merge branch 'main' of https://github.com/ruma/ruma into main

oppiliappan committed:
Merge branch 'main' into metadata-attributes

oppiliappan committed:
remove redundant return

jplatte commented:
GitHub is showing a commit from #400 as new in this PR. Could you rebase? (ask on Matrix if you need help with that)

jplatte commented on ruma-api-macros/src/api.rs:
```diff
@@ -83,8 +83,19 @@ pub fn expand_all(api: Api) -> syn::Result<TokenStream> {
+                rate_limited: #value,
+            }
+        })
+        .collect::<Vec<_>>();
```
You can collect this into a `TokenStream` directly (turbofish not needed if you annotate the binding), then you don't need the `#()` below.

```suggestion
        .collect();
```

jplatte commented on ruma-api-macros/src/api.rs:
```diff
@@ -227,19 +242,39 @@ pub fn expand_all(api: Api) -> syn::Result<TokenStream> {
-            impl #ruma_api::IncomingNonAuthRequest for #incoming_request_type {}
-        }
-    };
+    let authentication = &api
```
Any reason you moved this down here instead of keeping it at the top next to `rate_limited`? Otherwise, the same things apply.

oppiliappan replied:
I have moved it back up.

jplatte commented on ruma-api-macros/src/api.rs:
```diff
@@ -227,19 +242,39 @@ pub fn expand_all(api: Api) -> syn::Result<TokenStream> {
-                for Request #request_lifetimes
-            {}
+    let mut non_auth_endpoint_impls = vec![];
+    for auth in &api.metadata.authentication {
```
I think this can be simplified to

```rust
lte non_auth_endpoint_impls: TokenStream = api.metadata
	.authentication
    .map(|auth| if auth.value != "None" {
        TokenStream::new()
    } else {
        // ...
    })
    .collect();

jplatte commented on ruma-api-macros/src/api.rs:
```diff
@@ -83,8 +83,19 @@ pub fn expand_all(api: Api) -> syn::Result<TokenStream> {
     let path = &api.metadata.path;
-    let rate_limited = &api.metadata.rate_limited;
-    let authentication = &api.metadata.authentication;
+    let rate_limited = &api
```
```suggestion
    let rate_limited: TokenStream = api
```

jplatte commented on ruma-api-macros/src/api/metadata.rs:
```diff
@@ -54,6 +63,10 @@ fn set_field<T: ToTokens>(field: &mut Option<T>, value: T) -> syn::Result<()> {
     }
 }
 
+fn add_field<T: ToTokens>(field: &mut Vec<MetadataField<T>>, value: T, attrs: Vec<Attribute>) {
```
I don't think this needs to be a separate function, you can just inline it at the two usage sites.

jplatte commented on ruma-api-macros/src/api/metadata.rs:
```diff
@@ -91,8 +104,16 @@ impl Parse for Metadata {
+                Err(missing_field("rate_limited"))
+            } else {
+                Ok(rate_limited)
+            }?,
```
I think this would be a bit more readable as

```suggestion
            rate_limited: if rate_limited.is_empty() {
                return Err(missing_field("rate_limited"));
            } else {
                rate_limited
            },
```

jplatte commented on ruma-api-macros/src/api/metadata.rs:
```diff
@@ -139,12 +160,21 @@ enum FieldValue {
+            if !util::is_cfg_attribute(attr) {
+                return Err(syn::Error::new_spanned(
+                    &attr,
+                    "attributes may only be `cfg` attributes",
```
The way this is currently worded it sounds like a general statement, not specific to the `metadata` block. This is only a suggestion but I think including "here" in the message would be good.

```suggestion
                    "only `cfg` attributes may appear here",
```

oppiliappan replied:
> The way this is currently worded it sounds like a general statement

Agreed, using the word "attributes" twice didn't sit well with me either.

oppiliappan committed:
make changes as requested

 - collect quoted expansions into TokenStream
 - inline add_field
 - improve error message on non-cfg attributes

jplatte commented:
>  - [ ] add tests (I am not sure how this is done)

Just add different values for authentication and rate_limited under `cfg(all())` (always true) and `cfg(any())` (always false) to one of the tests in `ruma-api/src/ruma_api_macros.rs`.

>  - [ ] emit correct error message on duplicate field declaration

We can skip this. It would require somehow detecting when `cfg` expressions overlap, which would be too complicated.

oppiliappan committed:
update test to cover metadata attributes

jplatte commented:
Thanks!
